PINCache/PINDiskCache.m

```diff
@@ -143,7 +146,24 @@ - (NSString *)encodedString:(NSString *)string
 {
     if (![string length])
         return @"";
-   return [string stringByAddingPercentEncodingWithAllowedCharacters:[[NSCharacterSet characterSetWithCharactersInString:@".:/"] invertedSet]];
+    
+    NSString *encodedString = [string stringByAddingPercentEncodingWithAllowedCharacters:
+                               [[NSCharacterSet characterSetWithCharactersInString:@".:/"] invertedSet]];
+    
+    const char *fileSystemRepresentation = [encodedString fileSystemRepresentation];
+    size_t length = strlen(fileSystemRepresentation);
+    
+    if (length > NAME_MAX) {
+        //encoded string name will be too long for file representation, md5 and hope for no collisions.
+        unsigned char sha1Buffer[CC_SHA1_DIGEST_LENGTH];
+        CC_SHA1(fileSystemRepresentation, (CC_LONG)length, sha1Buffer);
+        encodedString = @"";
+        for (NSUInteger idx = 0; idx < CC_SHA1_DIGEST_LENGTH; idx++) {
+            encodedString = [encodedString stringByAppendingFormat:@"%02x", sha1Buffer[idx]];
```

Comment:
If you're going to build up a string like this, you'd be better using an `NSMutableString` or, even better, writing into a C character buffer on the stack (because you know the exact buffer length you need) and then converting that to an `NSString` for the return value.